### PR TITLE
[kml] Keep normal classifier types. Safe ser/des.

### PR DESCRIPTION
**Pull request by vng**

### CODE REVIEW

File: libs/map/bookmark_helpers.cpp [hunk 1]
```diff
@@ -733,38 +732,6 @@ BookmarkBaseType GetBookmarkBaseType(std::vector<uint32_t> const & featureTypes)
...
-  {
-    auto const itIcon = kFeatureTypeToBookmarkMatchInfo.find(typeStr);
-    if (itIcon != kFeatureTypeToBookmarkMatchInfo.cend())
-      return itIcon->second.m_icon;
-  }
-  while (TruncType(typeStr));
-
-  return kml::BookmarkIcon::None;
-}
-
```
Comment (biodranik) [inline libs/map/bookmark_helpers.cpp:750]: Build break on iOS and Android: `SaveFeatureTypes` and `GetBookmarkIconByFeatureType` are removed here but still called from:
- `android/sdk/src/main/cpp/app/organicmaps/sdk/bookmarks/data/BookmarkManager.cpp:298` — `SaveFeatureTypes(info.GetTypes(), bmData);`
- `iphone/Maps/UI/PlacePage/PlacePageManager/MWMPlacePageManager.mm:204` — `SaveFeatureTypes(info.GetTypes(), bmData);`
- `iphone/Maps/Core/Search/SearchResult.mm:120` — `GetBookmarkIconByFeatureType(featureType);`

Either keep these helpers (now operating on real types instead of indices) or update all three callers in the same commit. Since `bmData.m_featureTypes` semantics changed to store types, `SaveFeatureTypes` only needs `bmData.m_featureTypes.push_back(*it)` without the index conversion — trivial to adapt.

File: libs/kml/types.hpp [hunk 2]
```diff
@@ -340,7 +351,7 @@ struct BookmarkData
   LocalizableString m_description;
   // Bound feature's types: type indices sorted by importance, the most
   // important one goes first.
-  std::vector<uint32_t> m_featureTypes;
+  ClassifierTypes m_featureTypes;
```
Comment (biodranik) [inline libs/kml/types.hpp:354]: Runtime regression in `libs/map/framework.cpp:595-596` (`FillPointInfoForBookmark`):
```cpp
for (uint32_t i : bmk.GetData().m_featureTypes)
  types.push_back(cl.GetTypeForIndex(i));
```
After this PR, `m_featureTypes` stores real classifier types (not indices). Calling `GetTypeForIndex(type)` treats a 32-bit packed type as an index; since real types are much larger than `m_types.size()` (≈1647), this returns `INVALID_TYPE` (0) for virtually every bookmark. As a result, the equality lambda at framework.cpp:601 never matches, so the place-page selection never resolves to the underlying `FeatureType` for bookmarks that had bound features. The loop should just be `types.assign(bmk.GetData().m_featureTypes.begin(), bmk.GetData().m_featureTypes.end());`. Please include the framework.cpp fix in this PR.

Also update the stale comment on this member — it still says _"type indices sorted by importance"_ but they are no longer indices. Same in types_v3.hpp, types_v7.hpp, types_v8.hpp, types_v8mm.hpp.

File: libs/kml/visitors.hpp [hunk 3]
```diff
@@ -576,10 +592,25 @@ class BookmarkDeserializerVisitor
...
+  {
+    auto const sz = ReadVarUint<uint32_t>(m_source);
+    types.reserve(sz);
+    for (uint32_t i = 0; i < sz; ++i)
+    {
+      uint32_t index;
+      (*this)(index);
+      // Type can be removed after some time.
+      uint32_t const type = m_cl.GetTypeForIndex(index);
+      if (type != m_cl.GetStubType())
```
Comment (biodranik) [inline libs/kml/visitors.hpp]: This filter only catches the stub type but not `Classificator::INVALID_TYPE` (0). `GetTypeForIndex` returns `INVALID_TYPE` when `index >= m_types.size()` (e.g. truncated/corrupt data or a KMB written by a newer version with more types than the current build). Pushing 0 into `types` makes downstream `classif().GetReadableObjectName(0)` hit `ASSERT(!s.empty())` and then `pop_back()` on an empty string — UB in release builds.

```cpp
if (type != m_cl.GetStubType() && type != Classificator::INVALID_TYPE)
  types.push_back(type);
```

The XML-side counterpart (`serdes.cpp:1373`) already uses `type != Classificator::INVALID_TYPE`; converging on a single predicate is worth it.

File: libs/indexer/feature.cpp [hunk 4]
```diff
@@ -290,9 +290,12 @@ void FeatureType::ParseTypes()
   size_t const count = GetTypesCount();
   for (size_t i = 0; i < count; ++i)
   {
-    uint32_t index = ReadVarUint<uint32_t>(source);
+    /// @todo I suppose, better to skip Stub types at all, and keep only one if the result is empty.
+    /// AFAIR, tehre are no constraints for deserializable types count ..
```
Comment (biodranik) [inline libs/indexer/feature.cpp]: Nit: typo `tehre` → `there`. Also the `@todo` reads a bit awkwardly — consider:

```cpp
/// @todo Consider skipping Stub types entirely and retaining a single Stub only if the result would otherwise be empty.
/// There are no constraints on the count of deserialized types, AFAIR.
```

File: libs/kml/type_utils.hpp [hunk 5]
```diff
@@ -126,14 +126,22 @@ std::string GetPreferredBookmarkStr(LocalizableString const & name, feature::Reg
                                     std::string const & languageNorm);
 std::string GetLocalizedFeatureType(std::vector<uint32_t> const & types);
 
+// m_collectionIndex is mutable because it is filled when serialization.
```
Comment (biodranik) [inline libs/kml/type_utils.hpp]: Grammar nit: `because it is filled when serialization` → `because it is filled during serialization`.

File: libs/kml/types.hpp [hunk 6]
```diff
@@ -301,6 +301,17 @@ struct ColorData
   uint32_t m_rgba = 0;
 };
 
+class ClassifierTypes : public std::vector<uint32_t>
```
Comment (biodranik) [inline libs/kml/types.hpp:304]: Inheriting publicly from `std::vector` carries the usual footguns (non-virtual destructor, slicing through base pointers). Since the only value added here is a `DebugPrint` overload, two cheaper alternatives are worth considering:

1. Make `ClassifierTypes` a struct that holds `std::vector<uint32_t> m_types` — avoids the inheritance issue, but requires wrapping all vector operations.
2. Keep it a type alias (`using ClassifierTypes = std::vector<uint32_t>;`) and add the specialized visitor overloads by tag — but that loses compile-time distinction.

Public inheritance is fine here because we never delete through a base pointer and never slice — but a one-line `static_assert(std::is_standard_layout_v<ClassifierTypes>);` / comment documenting the intent would help future maintainers not worry about it.

File: libs/map/bookmark_helpers.cpp [hunk 7]
```diff
@@ -35,7 +34,7 @@ struct BookmarkMatchInfo
   BookmarkBaseType m_type;
 };
 
-std::map<std::string, BookmarkMatchInfo> const kFeatureTypeToBookmarkMatchInfo = {
+std::map<std::string_view, BookmarkMatchInfo> const kFeatureTypeToBookmarkMatchInfo = {
```
Comment (biodranik) [inline libs/map/bookmark_helpers.cpp]: Nit: since all keys are static string literals, a `base::SortedArray` / sorted `std::vector<std::pair<std::string_view, BookmarkMatchInfo>>` with `std::lower_bound` would beat `std::map` on both build time and lookup without any ergonomic cost. `std::unordered_map<std::string_view, …>` is also a strict win over `std::map` here. Not blocking.

File: libs/kml/visitors.hpp [hunk 8]
```diff
@@ -364,6 +370,13 @@ class BookmarkSerializerVisitor
 
   void operator()(BookmarkIcon icon, char const * /* name */ = nullptr) { (*this)(static_cast<uint16_t>(icon)); }
 
+  void operator()(ClassifierTypes const & types, char const * /* name */ = nullptr)
+  {
+    WriteVarUint(m_sink, static_cast<uint32_t>(types.size()));
+    for (uint32_t t : types)
```
Comment (biodranik) [inline libs/kml/visitors.hpp:388]: Since `m_cl` is already cached as a member, `GetIndexForType` is a direct call — fine. One simplification: `operator()` on `uint32_t` just calls `WriteVarUint(m_sink, v)`, so you can skip the self-call and write the VarUint directly to make the intent obvious:

```cpp
void operator()(ClassifierTypes const & types, char const * = nullptr)
{
  WriteVarUint(m_sink, static_cast<uint32_t>(types.size()));
  for (uint32_t t : types)
    WriteVarUint(m_sink, m_cl.GetIndexForType(t));
}
```

Also — if the whole point of keeping `ClassifierTypes` as a VarUint vector of *indices* on the wire is just this translation, a comment at the top of the overload stating that the on-disk format preserves index encoding (for backward/forward compatibility with existing KMB files) would save the next reader a diff dive.

File: libs/kml/visitors.hpp [hunk 9]
```diff
@@ -703,11 +734,10 @@ class DeserializedStringCollector
     if (!SwitchSubIndexIfNeeded(index))
       return;
 
-    // TODO(AB): A copy of map is created, is it really necessary?
-    auto subIndex = index[m_counter];
+    auto const & subIndex = index[m_counter];
     auto const sz = static_cast<int8_t>(subIndex.size() / 2);
```
Comment (biodranik) [inline libs/kml/visitors.hpp:754]: Switching from `subIndex[i]` to `subIndex.at(i)` is unrelated to the main refactor — fine as an opportunistic bounds-check hardening, but please split this into a separate commit (or at minimum call it out in the PR description). Note that `std::map::at` throws `std::out_of_range` on an unknown key, so an inconsistency between `sz` and the keys present in `subIndex` now terminates the whole parse instead of silently default-constructing — that's an intentional behavior change worth flagging.

**Review by biodranik — CHANGES_REQUESTED** [review]

File: libs/kml/visitors.hpp [hunk 10]
```diff
@@ -71,21 +73,21 @@ class CollectorVisitor
...
   }
 
   template <typename T>
-  std::enable_if_t<!VisitedTypes<T>::value> VisitIfPossible(T & t)
+  std::enable_if_t<!VisitedTypes<T>::value> VisitIfPossible(T && t)
   {}
 
   template <typename T>
-  void operator()(T & t, char const * /* name */ = nullptr)
+  void operator()(T && t, char const * /* name */ = nullptr)
```
Comment (biodranik) [inline libs/kml/visitors.hpp:93]: **Blocker — this change breaks both the const and non-const collector paths (verified by running `kml_tests`).**

Switching from `T &` to forwarding reference `T &&` causes `T` to be deduced as a *reference type* (e.g. `FileData &`, `BookmarkData const &`) for every lvalue argument. The trait predicates are applied to that reference type:

```cpp
HasCollectionMethods<BookmarkData &>::value      // -> 0  (can't form &C::ClearCollectionIndex when C is a reference)
VisitedTypes<BookmarkData &>::value               // -> 0  (is_same<BookmarkData &, BookmarkData> is false)
```

So both `PerformActionIfPossible` and `VisitIfPossible` pick the empty overload — `Collect()` is never called, nested `Visit()` is never invoked, and `ClearCollectionIndex()` is a no-op. Tests now fail:

```
serdes_tests.cpp::Kml_Deserialization_Text_Bin_Memory   [FAILED]  (collectionIndex left populated on dataFromBin)
serdes_tests.cpp::Kml_Serialization_Bin_Memory          [FAILED]  (binary payload diverges from golden)
serdes_tests.cpp::Kml_Serialization_Bin_File            [CRASH]   (downstream CHECK fires)
```

Fix options, in order of preference:

1. **Strip cv/ref in the traits** (minimal change, preserves the new const-correct signature) and also add a const overload for the vector traversal:
   ```cpp
   template <typename T>
   std::enable_if_t<HasCollectionMethods<std::remove_cvref_t<T>>::value>
   PerformActionIfPossible(T && t) { ... }

   template <typename T>
   std::enable_if_t<VisitedTypes<std::remove_cvref_t<T>>::value>
   VisitIfPossible(T && t) { ... }

   template <typename T>
   void operator()(std::vector<T> const & vs, char const * = nullptr)
   { for (auto const & v : vs) (*this)(v); }
   ```

2. **Revert to `T &`** and keep `SerializerKml` taking `FileData &` until the collector is made genuinely const-correct. This PR gains little from the const switch — the only mutation it avoids is `m_collectionIndex`, which the PR already makes `mutable`.

Either way, please add a round-trip test that serializes a BookmarkData with a non-default `m_name`/`m_description`/`m_properties` and asserts the round-tripped payload equals the original (byte-for-byte for `Kml_Serialization_Bin_Memory`).

File: libs/kml/types.hpp (hunk 6 above)
Comment (biodranik) [inline libs/kml/types.hpp:304]: Secondary observation on the regressions surfaced in this review: because `class ClassifierTypes : public std::vector<uint32_t>` is implicitly convertible from any `std::vector<uint32_t>`, stale call sites that still feed *indices* keep compiling cleanly even though the semantics flipped. Two confirmed casualties in-tree:

- `libs/map/framework.cpp:595` — still calls `GetTypeForIndex(i)` on what is now a real type (covered in the other comment on this file's `m_featureTypes`).
- `libs/kml/pykmlib/bindings.cpp:812` exposes `BookmarkData.feature_types` directly, and `bindings.cpp:643` / `bindings_test.py:48-50` feed the field with `classificator_type_to_index(...)`. On the next serialize, `BookmarkSerializerVisitor::operator()(ClassifierTypes const &)` calls `m_cl.GetIndexForType(small_index)`, which fails the CHECK in `IndexAndTypeMapping::GetIndex` and aborts the process. Please migrate the binding (add `classificator_readable_name_to_type` and update the test) or translate at the binding boundary.

A strong-typedef that required an explicit constructor — or even just an explicit `IndicesToTypes` / `TypesToIndices` helper invoked at the ser/des boundary rather than inside a shadow wrapper — would have let the compiler surface every broken call site in this same PR.

File: libs/kml/types.hpp (hunk 2 above)
Comment (vng) [reply libs/kml/types.hpp:354]: This will be the next, that is why I started all the work here.

File: libs/kml/visitors.hpp (hunk 8 above)
Comment (vng) [reply libs/kml/visitors.hpp:388]: No, keep the format - there is not WriteVarUint but WriteToSink

File: libs/kml/visitors.hpp (hunk 10 above)
Comment (vng) [reply libs/kml/visitors.hpp:93]: This will be fine, I forgot to add an overload.

File: libs/kml/visitors.hpp (hunk 9 above)
Comment (vng) [reply libs/kml/visitors.hpp:754]: Adding dummy indices {0, 0} or Extracting a string from {0, 0} or adding an empty (in the best scenario) string doesn't add stability here.

File: libs/indexer/feature_data.cpp [hunk 11]
```diff
@@ -202,6 +208,47 @@ void TypesHolder::SortBySpec()
...
+  size_t i = 0;
+  for (; i < Size() && i < other.Size(); ++i)
+  {
+    bool const ul = checker.IsUseful(m_types[i]);
+    if (ul != checker.IsUseful(other.m_types[i]))
+      return false;
+
+    if (!ul)
+    {
+      ASSERT(i > 0, ("Show VNG this case!", *this, other));
```
Comment (biodranik) [inline libs/indexer/feature_data.cpp:236]: Two problems:

1. The message `"Show VNG this case!"` is a personal debug marker — not something that should land in `master`.
2. Reaching `i == 0` with both sides non-useful is **reachable** in production: any pair of features whose top type (after `SortToCompare`) lands in `UselessTypesChecker::m_types` (e.g. a `building`-only pair) hits this branch. In release the assert is compiled out and the function silently returns `false`, hiding a legitimate match.

Suggest dropping the assert entirely; `return i > 0;` already encodes the intent, and the case stops being surprising:

```cpp
if (!ul)
  return i > 0;
```

File: libs/map/framework.cpp [hunk 12]
```diff
@@ -589,22 +589,20 @@ kml::MarkGroupId Framework::AddCategory(std::string const & categoryName)
...
-    if (types.empty() || ft.GetTypesCount() != types.size())
+    if (bmTypes.Empty())
       return false;
 
-    // Strict equal types.
     feature::TypesHolder fTypes(ft);
-    std::sort(fTypes.begin(), fTypes.end());
-    return std::equal(types.begin(), types.end(), fTypes.begin(), fTypes.end());
+    fTypes.SortToCompare();
+    return bmTypes.EqualUsefulSorted(fTypes);
```
Comment (biodranik) [inline libs/map/framework.cpp:605]: Matching regression worth discussing before merge.

`EqualUsefulSorted` returns `true` as soon as both sides reach the first non-useful type, so trailing useless tags (`cuisine`, `psurface`, `wheelchair`, `internet_access`, `fee`, …) are ignored. The matcher is plugged into `SelectionProcessor::FillPointInfo` (see `selection_processor.cpp:276`), which sets `result.m_poi` on the first `geometryHit && matcher(ft)`. Two colocated POIs sharing only their useful-type prefix — e.g. `amenity-cafe` + `cuisine=vegan` vs `amenity-cafe` + `cuisine=italian`, or `tourism-hotel` + different `wheelchair`/`internet_access` — can alias. The bookmark PP binds to whichever feature is visited first and can show the wrong metadata.

Suggest keeping strict equality as the first pass and only falling back to `EqualUsefulSorted` when no exact candidate exists:

```cpp
// First pass: strict equal on all types.
auto strictMatcher = [&bmTypes](FeatureType & ft) {
  feature::TypesHolder fTypes(ft);
  fTypes.SortToCompare();
  return bmTypes.Size() == fTypes.Size()
      && std::equal(bmTypes.begin(), bmTypes.end(), fTypes.begin());
};

// Fallback: useful-only.
auto lenientMatcher = [&bmTypes](FeatureType & ft) {
  if (bmTypes.Empty())
    return false;
  feature::TypesHolder fTypes(ft);
  fTypes.SortToCompare();
  return bmTypes.EqualUsefulSorted(fTypes);
};
```

The `SelectionProcessor` would need a way to report whether it matched, or the caller can run a probe pass and then fall back.

Separately on this same block:

1. `bmTypes.Empty()` inside the lambda is loop-invariant — `SelectionProcessor` calls it for every candidate. Hoist the check outside and early-return when empty.
2. `std::min(kMaxTypesCount, srcTypes.size())` silently truncates when a malformed/newer KMB has more than 8 types — your own `@todo` in `feature.cpp:293` notes the deserializer doesn't enforce the count. A `CHECK_LESS_OR_EQUAL(srcTypes.size(), feature::kMaxTypesCount, ())` (or dropping the `min` and letting `TypesHolder::Add`'s CHECK fire) would make the failure mode observable rather than silently corrupting matching.

File: iphone/Maps/Core/Search/SearchResult.mm [hunk 13]
```diff
@@ -116,8 +116,7 @@ - (instancetype)initWithResult:(search::Result const &)result itemType:(SearchIt
 
     if (result.GetResultType() == search::Result::Type::Feature)
     {
-      auto const featureType = result.GetFeatureType();
-      auto const bookmarkImage = GetBookmarkIconByFeatureType(featureType);
+      auto const bookmarkImage = GetBookmarkMatchInfo({result.GetFeatureType()}).m_icon;
```
Comment (biodranik) [inline iphone/Maps/Core/Search/SearchResult.mm]: This replaces the old single-type icon lookup with a temporary `std::vector`, so we now heap-allocate once per feature search result row. Non-blocking, but this is a hot UI path — worth keeping a `GetBookmarkMatchInfo(uint32_t)` / `GetBookmarkIconByFeatureType(uint32_t)` overload and having the `std::vector` variant delegate to it.

```cpp
BookmarkMatchInfo GetBookmarkMatchInfo(uint32_t featureType);
inline BookmarkMatchInfo GetBookmarkMatchInfo(std::vector<uint32_t> const & types)
{
  for (auto t : types)
    if (auto info = GetBookmarkMatchInfo(t); info.m_icon != kml::BookmarkIcon::None)
      return info;
  return {kml::BookmarkIcon::None, BookmarkBaseType::None};
}
```

File: libs/indexer/feature_data.hpp [hunk 14]
```diff
@@ -133,6 +133,10 @@ class TypesHolder
   /// Sort types by it's specification (more detailed type goes first). Should be used in client app.
   void SortBySpec();
 
+  /// Sort types for comparison. Useless in the end, the order is trict (functions above calls stable_sort).
```
Comment (biodranik) [inline libs/indexer/feature_data.hpp]: Grammar + typos (`trict` → `strict`, `Useless in the end` parses as "useless [the function] in the end", `calls` → `call`). Also the function's precondition isn't obvious from the name. Suggest:

```cpp
/// Sort types for comparison: useless types go to the end, tie-broken by type value
/// so the order is deterministic (plain std::sort, not stable_sort like the functions above).
void SortToCompare();

/// Precondition: both holders are pre-sorted by SortToCompare.
/// Returns true iff the 'useful' (Score < 2) prefix matches element-wise; trailing
/// useless types (cuisine, psurface, wheelchair, fee, ...) are ignored. See the
/// matching-aliasing caveat on framework.cpp's FillPointInfoForBookmark.
bool EqualUsefulSorted(TypesHolder const & other) const;
```

File: libs/kml/types.hpp (hunk 2 above)
Comment (biodranik) [inline libs/kml/types.hpp:355]: Stale doc comment (lines 352-353 above): you acknowledged in the earlier thread that this would come in a follow-up, but it is a trivial change and the confusion is right next to the semantic flip. Member no longer stores indices.

Suggest:

```cpp
// Bound feature's classifier types, sorted by importance (most important first).
// On-disk format (KMB) stores indices for backward compatibility; see
// BookmarkSerializerVisitor / BookmarkDeserializerVisitor in visitors.hpp.
ClassifierTypes m_featureTypes;
```

Same stale comment needs updating in:
- `libs/kml/types_v7.hpp:70-71`
- `libs/kml/types_v8.hpp:85-86`
- `libs/kml/types_v8mm.hpp:59-60`

File: libs/kml/types.hpp [hunk 15]
```diff
@@ -301,6 +301,17 @@ struct ColorData
...
+{
+public:
+  using std::vector<uint32_t>::vector;
+
+  friend std::string DebugPrint(ClassifierTypes const & types)
+  {
+    return DebugPrintSequence(types.begin(), types.end());
```
Comment (biodranik) [inline libs/kml/types.hpp:311]: Two nits on `ClassifierTypes`:

1. `DebugPrint` now prints raw `uint32_t` values, which are opaque packed-type numbers — uninformative in logs. Consider mapping through `classif().GetReadableObjectName(t)` so debug output shows readable type names (e.g. `"amenity-cafe, cuisine-vegan"`).
2. Re-surfacing @biodranik's concern on public inheritance from `std::vector`: since the class ships as-is, a short in-class comment would at least document the contract and warn the next author about the implicit-convertibility footgun (which already bit `framework.cpp` in the initial review):

```cpp
// Real classifier types (not indices) held in descending importance order.
// Inherits from std::vector only to add a typed DebugPrint overload; do NOT
// add virtual members or pass ClassifierTypes* through a std::vector<uint32_t>*.
class ClassifierTypes : public std::vector<uint32_t> { ... };
```

File: libs/kml/serdes.cpp [hunk 16]
```diff
@@ -1371,15 +1369,9 @@ void KmlParser::CharData(std::string & value)
...
-              MYTHROW(DeserializerKml::DeserializeException, ("Types mapping is not loaded."));
-            auto const type = c.GetTypeByReadableObjectName(value);
-            if (c.IsTypeValid(type))
-            {
-              auto const typeInd = c.GetIndexForType(type);
-              m_featureTypes.push_back(typeInd);
-            }
+            auto const type = classif().GetTypeByReadableObjectName(value);
+            if (type != Classificator::INVALID_TYPE)
+              m_featureTypes.push_back(type);
```
Comment (biodranik) [inline libs/kml/serdes.cpp:1375]: Filter diverges from the KMB path in `BookmarkDeserializerVisitor` (`visitors.hpp:738`), which now rejects both `INVALID_TYPE` **and** `GetStubType()`. Here only `INVALID_TYPE` is filtered, so a stale XML readable name that resolves to a stub slips through, and the next save will call `GetReadableObjectName(stub)`, which `ASSERT(!s.empty())`s and then `pop_back()`s an empty string (UB in release).

One-liner alignment:

```cpp
auto const & c = classif();
auto const type = c.GetTypeByReadableObjectName(value);
if (type != Classificator::INVALID_TYPE && type != c.GetStubType())
  m_featureTypes.push_back(type);
```

File: libs/map/bookmark_helpers.cpp [hunk 17]
```diff
@@ -705,64 +708,44 @@ void ResetIds(kml::FileData & kmlData)
...
-    {
-      bmData.m_featureTypes.push_back(c.GetIndexForType(*it));
-      if (bmData.m_icon == kml::BookmarkIcon::None)
-        bmData.m_icon = GetBookmarkIconByFeatureType(*it);
-    }
-  }
+  for (uint32_t t : copy)
+    bmData.m_featureTypes.push_back(t);
+
+  bmData.m_icon = GetBookmarkMatchInfo(bmData.m_featureTypes).m_icon;
```
Comment (biodranik) [inline libs/map/bookmark_helpers.cpp:745]: Quiet behavior change worth flagging: the old `SaveFeatureTypes` only assigned `m_icon` when it was `kml::BookmarkIcon::None`; this now always overwrites.

Both current callers (iOS `MWMPlacePageManager.mm:204`, Android `BookmarkManager.cpp:298`) pass a fresh `bmData`, so today it's a no-op. But a future caller that pre-sets `m_icon` (e.g. user-picked icon then backfills types) would silently lose the user's choice. Either preserve the guard:

```cpp
if (bmData.m_icon == kml::BookmarkIcon::None)
  bmData.m_icon = GetBookmarkMatchInfo(bmData.m_featureTypes).m_icon;
```

or document `'Callers must not pre-set m_icon; this function will overwrite it.'` on the declaration.

File: libs/kml/kml_tests/serdes_tests.cpp [hunk 18]
```diff
@@ -192,6 +193,7 @@ kml::FileData GenerateKmlFileDataForTrackWithTimestamps()
 // 1. Check text and binary deserialization from the prepared sources in memory.
 UNIT_TEST(Kml_Deserialization_Text_Bin_Memory)
 {
+  classificator::Load();
```
Comment (biodranik) [inline libs/kml/kml_tests/serdes_tests.cpp:196]: You fixed the `CollectorVisitor` const path after @biodranik flagged that tests passed silently while the const traversal was a no-op. Worth locking that in with an explicit regression test so a future ref/cvref slip is caught:

```cpp
UNIT_TEST(Kml_Serialization_Bin_Memory_Const)
{
  classificator::Load();
  kml::FileData const data = GenerateKmlFileData();  // note: const
  std::vector<uint8_t> buf;
  {
    MemWriter writer(buf);
    kml::binary::SerializerKml ser(data);
    ser.Serialize(writer);
  }
  TEST(!buf.empty(), ());

  // Round-trip.
  kml::FileData restored;
  MemReader reader(buf.data(), buf.size());
  kml::binary::DeserializerKml des(restored);
  des.Deserialize(reader);
  TEST_EQUAL(data, restored, ());
}
```

The old `T &&` bug silently produced an empty/wrong buffer; a test like this would have failed loudly.

File: libs/indexer/feature_data.cpp (hunk 11 above)
Comment (vng) [reply libs/indexer/feature_data.cpp:236]: - Useless only features should not pass generator stage
- "building" is not a useless type (added test)

File: libs/map/bookmark_helpers.cpp (hunk 17 above)
Comment (vng) [reply libs/map/bookmark_helpers.cpp:745]: Previous if check was because of enclosing for loop. 
Always assigning the icon is what is intended here.

File: libs/kml/kml_tests/serdes_tests.cpp (hunk 18 above)
Comment (vng) [reply libs/kml/kml_tests/serdes_tests.cpp:196]: Added const to the existing tests.

File: libs/kml/types.hpp (hunk 15 above)
Comment (vng) [reply libs/kml/types.hpp:311]: Inherit implementation here to avoid copy-paste. Inheritance doesn't automatically mean polymorphic use.
